rust/core-lib/src/plugins/manifest.rs

```diff
@@ -107,10 +111,132 @@ pub enum PluginScope {
     SingleInvocation,
 }
 
+#[derive(Debug, Clone, Serialize, Deserialize)]
+/// Represents a custom command provided by a plugin.
+pub struct Command {
+    /// Human readable title, for display in (for example) a menu.
+    pub title: String,
+    /// A short description of the command.
+    pub description: String,
+    /// Template of the command RPC as it should be sent to the plugin.
+    pub rpc_cmd: PlaceholderRpc,
+    /// A list of `CommandArgument`s, which the client should use to build the RPC.
+    pub args: Vec<CommandArgument>,
+}
+
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+/// A user provided argument to a plugin command.
+pub struct CommandArgument {
+    /// A human readable name for this argument, for use as placeholder
+    /// text or equivelant.
+    pub title: String,
+    /// A short (single sentence) description of this argument's use.
+    pub description: String,
+    pub key: String,
+    pub arg_type: ArgumentType,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    /// If `arg_type` is `Choice`, `options` must contain a list of options.
+    pub options: Option<Vec<ArgumentOption>>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub enum ArgumentType {
+    Number, Int, PosInt, Bool, String, Choice,
```

Comment:
What I mean is that serde has attributes that let you get the JSON representation you want while also using the correct Rust representation:

https://serde.rs/enum-representations.html

If you use the "internally tagged" representation you should be able to get a representation like

```js
{"title": "My Argument", /* ... */, "type": {"name": "choice", "choices": ["Blah", "Foo", "Bar"]}
```

Which I think is better in that it puts the options as part of the type, which they are, and working with it in statically typed languages will be way nicer.